swarm/storage/mru/resource.go

```diff
@@ -318,87 +321,119 @@ func (h *Handler) chunkSize() int64 {
+func (h *Handler) New(ctx context.Context, request *UpdateRequest) error {
 
 	// frequency 0 is invalid
-	if frequency == 0 {
-		return nil, nil, NewError(ErrInvalidValue, "frequency cannot be 0")
+	if request.frequency == 0 {
+		return NewError(ErrInvalidValue, "frequency cannot be 0")
 	}
 
 	// make sure name only contains ascii values
-	if !isSafeName(name) {
-		return nil, nil, NewError(ErrInvalidValue, fmt.Sprintf("invalid name: '%s'", name))
+	if !isSafeName(request.name) {
+		return NewError(ErrInvalidValue, fmt.Sprintf("invalid name: '%s'", request.name))
 	}
 
-	// get our blockheight at this time
-	currentblock, err := h.getBlock(ctx, name)
-	if err != nil {
-		return nil, nil, err
+	// make sure owner is set
+	var zeroAddr = common.Address{}
+	if request.ownerAddr == zeroAddr {
+		return NewError(ErrInvalidValue, "ownerAddr must be set to create a new metadata chunk")
 	}
 
+	// get the current time
+	if request.startTime == 0 {
+		request.startTime = h.getCurrentTime(ctx)
+	}
 	// create the meta chunk and store it in swarm
-	chunk := h.newMetaChunk(name, currentblock, frequency, ownerAddr)
+
+	chunk, metaHash := h.newMetaChunk(&request.resourceMetadata)
 	h.chunkStore.Put(chunk)
 
-	nameHash := ens.EnsNode(name)
-	log.Debug("new resource", "name", name, "key", nameHash, "startBlock", currentblock, "frequency", frequency, "owner", ownerAddr)
+	request.metaHash = metaHash
+	request.rootAddr = chunk.Addr
+	request.period = 1
+	request.version = 1
+
+	log.Debug("new resource", "name", request.name, "startBlock", request.startTime, "frequency", request.frequency, "owner", request.ownerAddr)
 
 	// create the internal index for the resource and populate it with the data of the first version
 	rsrc := &resource{
-		startBlock: currentblock,
-		frequency:  frequency,
-		name:       name,
-		nameHash:   nameHash,
-		updated:    time.Now(),
+		resourceUpdate: resourceUpdate{
+			rootAddr: chunk.Addr,
+		},
+		resourceMetadata: resourceMetadata{
+			name:      request.name,
+			startTime: request.startTime,
+			frequency: request.frequency,
+		},
+		updated: time.Now(),
+	}
+	copy(rsrc.ownerAddr[:], request.ownerAddr[:])
+	h.set(chunk.Addr, rsrc)
+
+	return nil
+}
+
+func (h *Handler) NewUpdateRequest(ctx context.Context, rootAddr storage.Address) (*UpdateRequest, error) {
+
+	if rootAddr == nil {
+		return nil, NewError(ErrInvalidValue, "rootAddr cannot be nil")
 	}
-	copy(rsrc.owner[:], ownerAddr[:])
-	h.set(chunk.Addr.Hex(), rsrc)
 
-	return chunk.Addr, rsrc, nil
+	rsrc, err := h.Load(rootAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	currentblock := h.getCurrentTime(ctx)
+
+	mru := new(UpdateRequest)
+	mru.period, err = getNextPeriod(rsrc.startTime, currentblock, rsrc.frequency)
+	if err != nil {
+		return nil, err
+	}
+	if _, err = h.lookup(rsrc, mru.period, 0, 0); err != nil {
+		return nil, err
+	}
+
+	if !rsrc.isSynced() {
```

Comment:
I wonder, should this state be possible here after `Load` and `lookup`?

Reply:
You re right. I have fixed this directly on top of #704.